### PR TITLE
Adding error summary when account fails to be created

### PR DESCRIPTION
**Pull request by daynew**
We received a ticket from a Code.org employee who could not finish creating a new Code.org account. Every time they tried, the page would seem to reload. On that backend, what is probably happening is that there is an error when creating the new User. We don't know the root cause of why the employee's account failed to get created, but it is an issue that we aren't displaying any sort of error message. This PR addresses the lack of information by adding an error summary at the top of the `<form>` which displays every error encountered. This should greatly help us debug similar issues in the future.

As an added bonus, it is also a WCAG 2.1 Accessibility (A11y) requirement to add an error summary at the top of the form. This makes it much easier for vision impaired users to quickly diagnose what the issue is.

* Adds error summary when a new Code.org account fails to be created.
* Adds validation that a `:user_type` / "Account Type" was selected so we print the error message "Account Type is required" rather than "Account Type is not included in the list". I think the "is required" error message makes more sense to students/teachers.

Below are some screenshots of what the user will see if they encounter an error when creating a new Code.org account.
Screenshot 2023-09-18 at 4 12 21 PM
Screenshot 2023-09-18 at 1 27 15 PM
Note that in the Japanese screenshot above, "Account Type" and "Please correct the following errors:" show up in English. These are existing English strings we have in other parts of our codebase so they will be automatically translated by Crowdin during the next I18n sync.

## Links
* [JIRA](https://codedotorg.atlassian.net/browse/P20-232)

## Testing story
* Manually tested on localhost.

### CODE REVIEW

**Review by carl-codeorg — APPROVED** [review]
LGTM! Do we know what the error was that Sean encountered in the original Zendesk ticket? The integration in JIRA isn't working for this ticket for me, so I can't see what he originally reported.

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -187,7 +187,9 @@ class User < ApplicationRecord
     TYPE_STUDENT = 'student'.freeze,
     TYPE_TEACHER = 'teacher'.freeze
   ].freeze
-  validates_inclusion_of :user_type, in: USER_TYPE_OPTIONS
+
+  validates_presence_of :user_type
+  validates_inclusion_of :user_type, in: USER_TYPE_OPTIONS, if: -> {user_type.present?}
```
Comment (artem-vavilov) [inline dashboard/app/models/user.rb]:
```suggestion
  validates_inclusion_of :user_type, in: USER_TYPE_OPTIONS, if: :user_type?
```

File: dashboard/app/views/devise/registrations/_finish_sign_up.html.haml [hunk 2]
```diff
@@ -6,12 +6,16 @@
 - country_code = location&.country_code.to_s.upcase
 -# 'RD' is the 'Reserved' country code for private IP's, i.e. localhost.
 - us_ip = ['US', 'RD'].include?(country_code)
-- @page_title = t('activerecord.attributes.user.finish_sign_up_header')
+- if @user.errors.any?
+  - @page_title = t('activerecord.attributes.user.finish_sign_up_header_error')
+- else
+  - @page_title = t('activerecord.attributes.user.finish_sign_up_header')
```
Comment (artem-vavilov) [inline dashboard/app/views/devise/registrations/_finish_sign_up.html.haml:12]: I would recommend moving this to a helper method like `UsersHelper.registation_page_title`
https://www.rubyguides.com/2020/01/rails-helpers/

**Review by artem-vavilov — APPROVED** [review]

**Comment by daynew** [comment]
> Do we know what the error was that Sean encountered in the original Zendesk ticket?

@carl-codeorg , no we don't. That's why I felt it is important we have an error summary with all the errors listed, not just the ones we expect.

File: dashboard/config/locales/en.yml [hunk 3]
```diff
@@ -66,13 +66,15 @@ en:
...
         finish_sign_up_header: 'Finish creating your account'
+        finish_sign_up_header_error: 'Please correct the following errors:'
         finish_sign_up_subheader: 'Fill out the following information to finish creating a Code.org account for <strong>%{email}</strong>.'
         finish_sign_up_subheader_markdown: 'Fill out the following information to finish creating a Code.org account for **%{email}**.'
         finish_sign_up_subheader_provider: 'Fill out the following information to finish creating a Code.org account with <strong>%{provider}</strong> for <strong>%{email}</strong>.'
         finish_sign_up_subheader_provider_markdown: 'Fill out the following information to finish creating a Code.org account with **%{provider}** for **%{email}**.'
         reset_password_token: "The reset password link you have followed"
         error:
           future: can't be in the future
+        user_type: "Account Type"
```
Comment (mgc1194) [inline dashboard/config/locales/en.yml]: Nit: Prefer single quotes for non-interpolated strings.
Also, I don't see where this string is being used.

Comment (wilkie) [reply dashboard/config/locales/en.yml]: It's magically applied when creating the form because it is a user attribute iirc.

Comment (daynew) [reply dashboard/config/locales/en.yml]: https://guides.rubyonrails.org/active_model_basics.html#translation

File: dashboard/app/views/devise/registrations/_finish_sign_up.html.haml (hunk 2 above)
Comment (daynew) [reply dashboard/app/views/devise/registrations/_finish_sign_up.html.haml:12]: Why do you recommend this? The docs you linked mentioned it is helpful for shared code, but this logic is only on used once.

Comment (artem-vavilov) [reply dashboard/app/views/devise/registrations/_finish_sign_up.html.haml:12]: It's just a bad practice to keep business logic in views: https://patrickkarsh.medium.com/importance-of-avoiding-logic-in-views-ruby-on-rails-basics-6b53f976d2f1

Comment (artem-vavilov) [reply dashboard/app/views/devise/registrations/_finish_sign_up.html.haml:12]: It's not a blocker, just a recommendation

File: dashboard/config/locales/en.yml (hunk 3 above)
Comment (artem-vavilov) [reply dashboard/config/locales/en.yml]: @mgc1194 you can also check these: 
1.  https://guides.rubyonrails.org/i18n.html#translations-for-active-record-models
2. https://kapeli.com/cheat_sheets/Rails_i18n.docset/Contents/Resources/Documents/index

**Review by mgc1194 — APPROVED** [review]

**Pull request merged**